### ISSUE TITLE
Refactor Function::map

### ISSUE THREAD

**Issue by jgillis** [casadi/casadi#1476]
The current design allows you to call a function with different apple-pear pairs:

```
[ [apple0 , pear0], [apple1, pear1],[apple2, pear2],...]
```

This does not scale very well to large number of pairs.
The following may be better:

```
[ horzcat(apple0,apple1,apple2,...) , horzcat(pear0,pear2,pear2,...) ] 
```

**Comment by jaeandersson** [comment]
:+1

**Comment by ghorn** [comment]
why does it not scale well?

**Comment by jgillis** [comment]
Suppose you have a million apple-pear pairs.
In the current design you'd end up with millions of nodes in the expression graph

**Comment by ghorn** [comment]
instead of 2 nodes, each of dimension millions?

**Comment by jgillis** [comment]
Yes

**Comment by jaeandersson** [comment]
Another way of thinking about it is that currently, the memory overhead for representing the map grows linearly with the number of repeats. You can see this in the C-codegen, where the step of collecting the arguments can be very long. By grouping like this, only the number of right-hand-sides of the matrices will grow.

**Comment by jgillis** [comment]
Merged with #1494 .
